File: src/array_mgmt/array_manager.cpp
/*
 *   BSD LICENSE
 *   Copyright (c) 2021 Samsung Electronics Corporation
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or without
 *   modification, are permitted provided that the following conditions
 *   are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *     * Neither the name of Intel Corporation nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *   OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "array_manager.h"

#include <list>

#include "src/device/device_manager.h"
#include "src/include/pos_event_id.h"
#include "src/logger/logger.h"
#include "src/mbr/abr_manager.h"

namespace pos
{
ArrayManager::ArrayManager()
{
    arrayRebuilder = new ArrayRebuilder(this);
    DeviceManagerSingleton::Instance()->SetDeviceEventCallback(this);
    abrManager = new AbrManager();
}

ArrayManager::~ArrayManager()
{
    delete arrayRebuilder;
    delete abrManager;
}

int
ArrayManager::Create(string name, DeviceSet<string> devs, string raidtype)
{
    if (_FindArray(name) != nullptr)
    {
        return (int)POS_EVENT_ID::ARRAY_ALREADY_EXIST;
    }

    if (arrayList.size() >= ArrayMgmtPolicy::MAX_ARRAY_CNT)
    {
        POS_TRACE_DEBUG((int)POS_EVENT_ID::ARRAY_CNT_EXCEEDED,
            "ArrayManager cnt exceeded. current: {}", arrayList.size());
        return (int)POS_EVENT_ID::ARRAY_CNT_EXCEEDED;
    }

    ArrayComponents* array = new ArrayComponents(name, arrayRebuilder, abrManager);
    int ret = array->Create(devs, raidtype);
    if (ret == (int)POS_EVENT_ID::SUCCESS)
    {
        arrayList.emplace(name, array);
    }
    else
    {
        delete array;
    }
    return ret;
}

int
ArrayManager::Delete(string name)
{
    ArrayComponents* array = _FindArray(name);
    if (array == nullptr)
    {
        return (int)POS_EVENT_ID::ARRAY_WRONG_NAME;
    }

    int ret = array->Delete();
    if (ret == (int)POS_EVENT_ID::SUCCESS)
    {
        delete array;
        _Erase(name);
    }

    return ret;
}

int
ArrayManager::Mount(string name)
{
    ArrayComponents* array = _FindArray(name);
    if (array != nullptr)
    {
        return array->Mount();
    }

    return (int)POS_EVENT_ID::ARRAY_WRONG_NAME;
}

int
ArrayManager::Unmount(string name)
{
    ArrayComponents* array = _FindArray(name);
    if (array != nullptr)
    {
        return array->Unmount();
    }

    return (int)POS_EVENT_ID::ARRAY_WRONG_NAME;
}

int
ArrayManager::AddDevice(string name, string dev)
{
    ArrayComponents* array = _FindArray(name);
    if (array != nullptr)
    {
        return array->GetArray()->AddSpare(dev);
    }

    return (int)POS_EVENT_ID::ARRAY_WRONG_NAME;
}

int
ArrayManager::RemoveDevice(string name, string dev)
{
    ArrayComponents* array = _FindArray(name);
    if (array != nullptr)
    {
        return array->GetArray()->RemoveSpare(dev);
    }

    return (int)POS_EVENT_ID::ARRAY_WRONG_NAME;
}

int
ArrayManager::DeviceDetached(UblockSharedPtr dev)
{
    // TODO_MULTIARRAY:Finding an array containing the device
    ArrayComponents* array = _FindArray("");
    if (array != nullptr)
    {
        return array->GetArray()->DetachDevice(dev);
    }

    return 0;
}

bool
ArrayManager::ArrayExists(string name)
{
    return _FindArray(name) != nullptr;
}

int
ArrayManager::GetAbrList(std::vector<ArrayBootRecord>& abrList)
{
    int result = abrManager->GetAbrList(abrList);
    return result;
}

IArrayInfo*
ArrayManager::GetArrayInfo(string name)
{
    ArrayComponents* array = _FindArray(name);
    if (array != nullptr)
    {
        return array->GetArray();
    }
    else
    {
        return nullptr;
    }
}

int
ArrayManager::PrepareRebuild(string name)
{
    ArrayComponents* array = _FindArray(name);
    if (array == nullptr)
    {
        return (int)POS_EVENT_ID::ARRAY_WRONG_NAME;
    }

    return array->PrepareRebuild();
}

void
ArrayManager::RebuildDone(string name)
{
    ArrayComponents* array = _FindArray(name);
    if (array != nullptr)
    {
        array->RebuildDone();
    }
}

int
ArrayManager::Load(list<string>& failedArrayList)
{
    std::vector<ArrayBootRecord> abrList;
    std::vector<ArrayBootRecord>::iterator it;
    int result = abrManager->GetAbrList(abrList);
    int loadResult = 0;
    if (result == 0)
    {
        for (it = abrList.begin(); it != abrList.end(); it++)
        {
            loadResult = _Load(it->arrayName);
            if (loadResult != 0)
            {
                string arrayName(it->arrayName);
                POS_TRACE_ERROR(loadResult, "Array " + arrayName + " load failed");
                failedArrayList.push_back(arrayName);
                result = loadResult;
            }
        }
    }
    else
    {
        POS_TRACE_ERROR(result, "Failed to get abr list");
    }

    return result;
}

int
ArrayManager::_Load(string name)
{
    // TODO_MULTIARRAY : load all arrays automatically.

    ArrayComponents* array = _FindArray(name);
    if (array != nullptr)
    {
        return (int)POS_EVENT_ID::ARRAY_ALREADY_EXIST;
    }

    array = new ArrayComponents(name, arrayRebuilder, abrManager);
    int ret = array->Load();
    if (ret == (int)POS_EVENT_ID::SUCCESS)
    {
        arrayList.emplace(name, array);
    }
    else
    {
        delete array;
    }

    return ret;
}

int
ArrayManager::ResetMbr(void)
{
    int result = 0;
    int deleteResult = 0;
    for (auto iter : arrayList)
    {
        ArrayComponents* array = _FindArray(iter.first);
        if (array != nullptr)
        {
            result = array->GetArray()->CheckDeletable();
            if (result != 0)
            {
                POS_TRACE_WARN(deleteResult, "Cannot delete array " + iter.first);
                return result;
            }
        }
        else
        {
            result = (int)POS_EVENT_ID::ARRAY_STATE_NOT_EXIST;
            POS_TRACE_ERROR(result, "Cannot find array " + iter.first);
            return result;
        }
    }

    for (auto iter : arrayList)
    {
        deleteResult = Delete(iter.first);
        if (deleteResult == 0)
        {
            _Erase(iter.first);
        }
        else
        {
            POS_TRACE_ERROR(deleteResult, "Failed to delete array " + iter.first);
            result = deleteResult;
        }
    }

    if (result == 0)
    {
        result = abrManager->ResetMbr();
    }

    return result;
}

ArrayComponents*
ArrayManager::_FindArray(string name)
{
    // TODO_MULTIARRAY : for compatibility
    if (name == "" && arrayList.size() == 1)
    {
        return arrayList.begin()->second;
    }
    auto it = arrayList.find(name);
    if (it == arrayList.end())
    {
        return nullptr;
    }

    return it->second;
}

ArrayDevice*
ArrayManager::_FindDevice(string devSn)
{
    for (auto iter : arrayList)
    {
        IArrayDevice* dev = iter.second->GetArray()->FindDevice(devSn);
        if (dev != nullptr)
        {
            return static_cast<ArrayDevice*>(dev);
        }
    }
    return nullptr;
}

void
ArrayManager::_Erase(string name)
{
    if (name == "" && arrayList.size() == 1)
    {
        arrayList.clear();
    }
    else
    {
        arrayList.erase(name);
    }
}

} // namespace pos

File: src/array_mgmt/array_manager.rs
use std::borrow::Borrow;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use lazy_static::lazy_static;
use log::{info, warn, error};
use crate::array_components::array_components::ArrayComponents;
use crate::array_models::dto::device_set::DeviceSet;
use crate::include::array_mgmt_policy;
use crate::mbr::abr_manager::AbrManager;
use crate::mbr::mbr_info::ArrayBootRecord;
use crate::array::meta::array_meta::ArrayMeta;
use crate::include::pos_event_id::PosEventId;

lazy_static!{
    pub static ref ArrayManagerSingleton: Arc<Mutex<ArrayManager>> = {
        let array_manager = ArrayManager::new();
        info!("ArrayManager has been created");
        Arc::new(Mutex::new(array_manager))
    };
}

pub struct ArrayManager {
    arrayList: HashMap::<String, ArrayComponents>,
    abrManager: AbrManager,
}

impl ArrayManager {
    fn new() -> ArrayManager {
        ArrayManager {
            arrayList: HashMap::new(),
            abrManager: AbrManager::new(),
        }
    }

    pub fn Create(&mut self, name: String, devs: DeviceSet<String>, metaFt: String, dataFt: String) -> Result<(), PosEventId>{
        info!("Creating an array {} with devices {:?} with meta {} and data {}", name, devs, metaFt, dataFt);

        if self._FindArray(&name).is_some() == true {
            let eventId = PosEventId::CREATE_ARRAY_SAME_ARRAY_NAME_EXISTS;
            warn!("[{}] name duplicated: {name}", eventId.to_string());
            return Err(eventId);
        }

        if self.arrayList.len() >= array_mgmt_policy::MAX_ARRAY_CNT {
            let len = self.arrayList.len();

            let eventId = PosEventId::CREATE_ARRAY_EXCEED_MAX_NUM_OF_ARRAYS;
            warn!("[{}] Current num of arrays: {len}", eventId.to_string());

            return Err(eventId);
        }

        let mut components = ArrayComponents::new();
        // TODO: hands over abr manager
        components.Create(name.clone(), devs, metaFt, dataFt)?;
        self.arrayList.insert(name.clone(), components);

        Ok(())
    }

    pub fn Delete(&mut self, name: String) -> Result<(), PosEventId> {
        let array = match self._FindArray(&name) {
            Some(a) => a,
            None => {
                if self.AbrExists(&name) {
                    self._DeleteFaultArray(&name);
                }

                let eventId = PosEventId::DELETE_ARRAY_ARRAY_NAME_DOES_NOT_EXIST;
                warn!("[{}] array_name: {name}", eventId.to_string());
                return Err(eventId);
            }
        };

        array.Delete()?;
        self.arrayList.remove(&name);

        Ok(())
    }

    fn _FindArray(&mut self, name: &String) -> Option<&mut ArrayComponents> {
        self.arrayList.get_mut(name)
    }

    pub fn AbrExists(&self, name: &String) -> bool {
        let abrList = match self.GetAbrList() {
            Ok(list) => list,
            Err(e) => {
                error!("Failed to get abr list, {e}");
                return false;
            }
        };

        abrList.iter().find(|&abr| &std::str::from_utf8(&abr.arrayName).unwrap() == name).is_some()
    }

    pub fn GetAbrList(&self) -> Result<Vec::<ArrayBootRecord>, PosEventId>{
        self.abrManager.GetAbrList()
    }

    fn _DeleteFaultArray(&self, name: &String) -> Result<(), PosEventId> {
        self.abrManager.LoadAbr(name)?;
        self.abrManager.DeleteAbr(name)?;

        Ok(())
    }
}
